@here/harp-mapview/lib/VisibleTileSet.ts

```diff
@@ -831,107 +886,141 @@ export class VisibleTileSet {
         }
     }
 
-    // Computes the visible tiles for each supplied datasource.
-    private getVisibleTilesForDataSources(
-        worldCenter: THREE.Vector3,
+    private getVisibleTilesForDataSource(
+        dataSource: DataSource,
         zoomLevel: number,
-        dataSources: DataSource[],
+        rootTiles: TileKeyEntry[],
+        frustumIntersection: FrustumIntersection,
         elevationRangeSource: ElevationRangeSource | undefined
-    ): {
-        tiles: Array<{ dataSource: DataSource; visibleTiles: TileKeyEntry[] }>;
-        allBoundingBoxesFinal: boolean;
-    } {
-        const tiles = [];
-        let allBoundingBoxesFinal: boolean = true;
-
-        for (const dataSource of dataSources) {
-            const displayZoomLevel = dataSource.getDisplayZoomLevel(zoomLevel);
-
-            const tilingScheme = dataSource.getTilingScheme();
-            const useElevationRangeSource: boolean =
-                elevationRangeSource !== undefined &&
-                elevationRangeSource.getTilingScheme() === tilingScheme;
-
-            const tileBounds = new THREE.Box3();
-            const workList: TileKeyEntry[] = this.getRequiredInitialRootTileKeys(worldCenter);
+    ): { visibleTiles: TileKeyEntry[]; bboxesFinal: boolean } {
+        const tilingScheme = dataSource.getTilingScheme();
+        const useElevationRangeSource: boolean =
+            elevationRangeSource !== undefined &&
+            elevationRangeSource.getTilingScheme() === tilingScheme;
+
+        const displayZoomLevel = dataSource.getDisplayZoomLevel(zoomLevel);
+        const tileBounds = new THREE.Box3();
+        const workList = [...rootTiles];
+        const visibleTiles: TileKeyEntry[] = [];
+        let bboxesFinal: boolean = true;
+
+        while (workList.length > 0) {
+            const tileEntry = workList.pop();
+
+            if (tileEntry === undefined) {
+                continue;
+            }
 
-            const visibleTiles: TileKeyEntry[] = [];
+            const tileKey = tileEntry.tileKey;
+            const uniqueKey = TileOffsetUtils.getKeyForTileKeyAndOffset(tileKey, tileEntry.offset);
+            const cachedTileEntry = frustumIntersection.tiles.get(uniqueKey);
 
-            const tileFrustumIntersectionCache = this.createIntersectionCache(workList);
+            assert(cachedTileEntry !== undefined);
+            assert(cachedTileEntry!.area > 0);
 
-            while (workList.length > 0) {
-                const tileEntry = workList.pop();
+            if (tileKey.level > displayZoomLevel) {
```

Comment:
This is broken in very obscure way.
`displayZoomLevel` is actually bounded by `maxZoomLevel` of first datasource that happens be processes by this method.

If for example, this first  datasource allows only data up to zoomLevel 15, all other datasources (that have data up to, 20) will also draw tiles up to 15.

So, one we should calculate like this:
```
maxDisplayZoomLevel = enabledDataSources.map((ds) => ds.getDisplayZoomLevel(zoomLevel)).reduce(max, 0);
```
and it should be done in `getVisibleTilesForDataSources` where we know all the DSs to be processed.

Reply:
done.